src/ui.rs

```diff
@@ -32,6 +34,12 @@ pub struct Ui {
     output: UiOutput,
 }
 
+impl Debug for Ui {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Ui").finish_non_exhaustive()
+    }
+}
```

Comment:
nit: can `#[derive(Debug)]`

I think `instrument(skip(ui))` is better as we wouldn't be interested in `Ui`.